sdk/identity/src/token_credentials/powershell_credential.rs

```diff
@@ -0,0 +1,179 @@
+        Ok(_) => PathBuf::from_str("pwsh.exe"),
+        Err(_) => PathBuf::from_str("powershell.exe"),
+    }
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_default_powershell_path() -> PathBuf {
+    PathBuf::from_str("pwsh").unwrap()
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+struct PowerShellTokenResponse {
+    pub token: AccessToken,
+    pub expires_on: DateTime<Utc>,
+}
+
+/// Enables authenticating with Azure Active Directory using Azure PowerShell.
+///
+/// To use this credential:
+/// - Install the Azure PowerShell Module
+///   `Install-Module -Name Az -Scope CurrentUser -Repository PSGallery -Force`
+/// - Ensure that you are signed in using the `Connect-AzAccount` cmdlet.
+pub struct AzurePowerShellCredential {
+    powershell_path: Option<PathBuf>,
+}
+
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error)]
+pub enum AzurePowerShellCredentialError {
+    #[error("Failed to import the Az.Account module (is it installed?): {0}")]
+    ImportFailed(String),
+    #[error("Failed to authenticate using the Az.Account module (are you logged in using `Connect-AzAccount`?): {0}")]
+    CommandFailed(String),
+    #[error("PowerShell is not installed and available on the system PATH: {0}")]
+    PowerShellNotInstalled(Error),
+    #[error("Az.Account token response was not UTF-8 encoded")]
+    ResponseNotUtf8(Utf8Error),
+    #[error("Failed to deserialize Az.Account token response")]
+    ResponseFailedToDeserialize(serde_json::Error),
+    #[error("Unknown error of kind: {0:?}")]
+    UnknownError(ErrorKind),
+}
+
+fn error_to_powershellcredential_error<T>(
+    error: Error,
+) -> Result<T, AzurePowerShellCredentialError> {
+    match error.kind() {
+        ErrorKind::NotFound => Err(AzurePowerShellCredentialError::PowerShellNotInstalled(
+            error,
+        )),
+        error_kind => Err(AzurePowerShellCredentialError::UnknownError(error_kind)),
+    }
+}
+
+impl AzurePowerShellCredential {
+    pub fn new(powershell_path: Option<PathBuf>) -> AzurePowerShellCredential {
+        AzurePowerShellCredential {
+            powershell_path: powershell_path,
+        }
+    }
+
+    async fn get_access_token(
+        &self,
+        resource: Option<&str>,
+    ) -> Result<PowerShellTokenResponse, AzurePowerShellCredentialError> {
+        let pwsh = self
+            .powershell_path
+            .as_ref()
+            .unwrap_or(&*POWERSHELL_PATH)
+            .as_os_str();
+
+        match Command::new(pwsh)
+            .args(vec![
+                "-Command",
+                "Import-Module Az.Accounts -MinimumVersion 2.2.0 -PassThru",
+            ])
+            .output()
+            .await
+        {
+            Ok(response) if !response.status.success() => {
+                let output = String::from_utf8_lossy(&response.stderr);
+                return Err(AzurePowerShellCredentialError::ImportFailed(
+                    output.to_string(),
```

Comment:
the call to `to_string` is not necessary

Reply:
Hm, `from_utf8_lossy` returns a `Cow<str>`. I think the conversion to String is required here and FWIW AzureCliCredential does the same in this case. The other option is that maybe I don't really need to use lossy conversion in this case, but figured it was simpler to do that than handle an Err from potentially invalid UTF-8 in the error output of the command.

Reply:
Ah yes - sorry. `to_owned` might be clearer (and potentially every so slightly more performant), but you don't have to change it if you don't want.